### DISCUSSION TITLE
Comparing two images to see if they're identical

### DISCUSSION THREAD

**Discussion by tfreedman** [ImageMagick/ImageMagick#5699] [Help]
I have a bunch of uncompressed .TIFF files from a film scanner, and I used Adobe Lightroom to losslessly convert them to DNG. I wanted to confirm that Lightroom created pixel-perfect conversions before deleting the originals, so I tried to compare them using a technique that has worked fairly well for me in the past:

convert a.tiff a.mpc
shasum -a 256 a.cache

convert b.dng b.mpc
shasum -b 256 b.cache

Unfortunately, the hashes don't match, suggesting that something changed along the way. I assume this is as simple as a color profile being added to the DNG file. Is there a better way to compare these files and verify that the pixel data is exactly identical, or otherwise determine what changed?

**Comment by fmw42** [comment]
Do a pixel by pixel comparison.

`compare -metric rmse a.tiff b.dng null:`

The lower the values the better. A perfect match would be

0 (0)

The first number is the value in the range 0 to quantumrange (typically 65535). The second number is the value in the range 0 to 1. (Multiply the second number by 100 to get a percent-like result).

**Comment by tfreedman** [comment]
That command gives me 15421.3 (0.235314)% as output on one set of images, and 21090.1 (0.321814)% on a new pair of images that I just converted.

**Reply by fmw42** [reply]
That is a large difference.  It should be visible if you flicker the two images in the display quickly.  Are you getting a proper DNG file from the TIFF?  Are you looking at the DNG JPG preview rather than the raw DNG data?

**Reply by tfreedman** [reply]
I should be looking at the raw DNG data (that was one of my concerns from earlier). I can see a very slight difference in Photoshop when they're subtracted from each other, but it isn't clear what's causing it (both sets involved Lightroom's lossless conversion mode, and I'm assuming that Adobe's developers didn't accidentally introduce a bug that would modify photos you intended to preserve). I would assume the pixel data is identical, unless something stupid is happening (like assuming a different color space in one file format vs another).

**Reply by fmw42** [reply]
Is DNG applying automatic brightness control and other miscellaneous processing that is not in the TIFF?

**Reply by tfreedman** [reply]
That's what I'd like to find out

**Reply by fmw42** [reply]
See DNG -defines and use them in your command to turn off any that might be one and changing your image.  https://imagemagick.org/script/defines.php

**Reply by tfreedman** [reply]
I can see certain programs naively interpreting either file format incorrectly, pre- and post-conversion. Since these are film scans, they don't have a native color space (and they're wonky to begin with due to discoloration). When converting to DNG, I can see someone assuming certain defaults about the file format when reading the image data.

**Reply by fmw42** [reply]
Since you are using Lightroom to make your DNG, you should consult with the Lightroom docs and forums. The DNG controls are then not really an Imagemagick issue here, since you are not using Imagemagick to create the DNG.  

Your compare values are showing 25 and 30 % errors. That should be dramatic if you look at the two results, not a small difference.

Post your two images and we can examine them to see what else might be different or post the results from identify -verbose on each one and do a diff to see where there are differences.

**Comment by snibgo** [comment]
Converting from TIFF to DNG could be a trivial change to metadata (because a DNG file *is* a TIFF file), but Lightroom may be making other changes, eg converting from non-linear sRGB to linear RGB. This will give you more information:
```
magick {your_file} -verbose info:
```

**Reply by fmw42** [reply]
Same as

identify -verbose {your_file}

**Comment by tfreedman** [comment]
Here are two files that exhibit this problem. They aren't the exact files I was working with earlier (those are around half a GB each on average), so I shrunk one down in Photoshop to 10%, exported it as TIFF, then imported it into Lightroom to do the DNG conversion. The original file's EXIF data is below, just in case something is different between that and the scaled down version.

exiftool -a -G /img20220907_17521610.tif
```
[ExifTool]      ExifTool Version Number         : 12.26
[File]          File Name                       : img20220907_17521610.tif
[File]          File Size                       : 184 MiB
[File]          File Modification Date/Time     : 2022:09:07 17:52:16-04:00
[File]          File Access Date/Time           : 2022:09:20 05:50:27-04:00
[File]          File Inode Change Date/Time     : 2022:09:07 17:52:16-04:00
[File]          File Permissions                : -rwx------
[File]          File Type                       : TIFF
[File]          File Type Extension             : tif
[File]          MIME Type                       : image/tiff
[File]          Exif Byte Order                 : Little-endian (Intel, II)
[EXIF]          Image Width                     : 5669
[EXIF]          Image Height                    : 5661
[EXIF]          Bits Per Sample                 : 16 16 16
[EXIF]          Compression                     : Uncompressed
[EXIF]          Photometric Interpretation      : RGB
[EXIF]          Strip Offsets                   : (Binary data 53338 bytes, use -b option to extract)
[EXIF]          Orientation                     : Horizontal (normal)
[EXIF]          Samples Per Pixel               : 3
[EXIF]          Rows Per Strip                  : 1
[EXIF]          Strip Byte Counts               : (Binary data 33965 bytes, use -b option to extract)
[EXIF]          X Resolution                    : 4800
[EXIF]          Y Resolution                    : 4800
[EXIF]          Planar Configuration            : Chunky
[EXIF]          Resolution Unit                 : inches
[Composite]     Image Size                      : 5669x5661
[Composite]     Megapixels                      : 32.1
```
identify -verbose /126/img20220907_17521610.tif
```
Image:
  Filename: /126/img20220907_17521610.tif
  Format: TIFF (Tagged Image File Format)
  Mime type: image/tiff
  Class: DirectClass
  Geometry: 5669x5661+0+0
  Resolution: 4800x4800
  Print size: 1.18104x1.17938
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: TrueColor
  Endianness: LSB
  Depth: 16-bit
  Channel depth:
    Red: 16-bit
    Green: 16-bit
    Blue: 16-bit
  Channel statistics:
    Pixels: 32092209
    Red:
      min: 2756  (0.0420539)
      max: 65535 (1)
      mean: 37050.2 (0.56535)
      median: 65535 (1)
      standard deviation: 15505.3 (0.236596)
      kurtosis: -0.630517
      skewness: -0.32685
      entropy: 0.932591
    Green:
      min: 1848  (0.0281987)
      max: 65535 (1)
      mean: 39860 (0.608225)
      median: 65535 (1)
      standard deviation: 15194 (0.231845)
      kurtosis: -0.187175
      skewness: -0.723296
      entropy: 0.916869
    Blue:
      min: 421  (0.00642405)
      max: 65535 (1)
      mean: 37693.9 (0.575172)
      median: 65535 (1)
      standard deviation: 13972.9 (0.213212)
      kurtosis: 0.422505
      skewness: -0.679648
      entropy: 0.883998
  Image statistics:
    Overall:
      min: 421  (0.00642405)
      max: 65535 (1)
      mean: 38201.4 (0.582916)
      median: 65535 (1)
      standard deviation: 14890.7 (0.227218)
      kurtosis: -0.22754
      skewness: -0.556783
      entropy: 0.911153
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 5669x5661+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: None
  Orientation: TopLeft
  Properties:
    date:create: 2022-09-07T21:52:16+00:00
    date:modify: 2022-09-07T21:52:16+00:00
    signature: 5bac5329714ca28bdbb0c9c493b5c343d31480b4d93dfa26735c6872b81c0e9c
    tiff:alpha: unspecified
    tiff:endian: lsb
    tiff:photometric: RGB
    tiff:rows-per-strip: 1
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 183.676MiB
  Number pixels: 32.0922M
  Pixels per second: 3399600P
  User time: 0.610u
  Elapsed time: 0:10.439
  Version: ImageMagick 7.1.0-20 Q16-HDRI x86_64 2022-01-22 https://imagemagick.org
```
[Archive.zip](https://github.com/ImageMagick/ImageMagick/files/9883343/Archive.zip)

**Reply by fmw42** [reply]
Do the same for your DNG file and look at the differences between the two results from identify -verbose. You will see the two are different.  I also note that your DNG seems to be doing some white balance (wb) (and red balance, rb) as I see

    dng:wb.rb.levels: 0.000000 0.000000 1.000000 0.000000

Though I do not know if this is just a null default as I am not a DNG expert.

If I view both files, I see that the clouds in the DNG file seem to more pink than in the tiff. So there seems to be some differences in coloration.
